src/Algebra/Lattice/Flat.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE CPP                #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveFoldable     #-}
+{-# LANGUAGE DeriveFunctor      #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DeriveTraversable  #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE TypeOperators      #-}
+#if __GLASGOW_HASKELL__ < 709
+{-# LANGUAGE Trustworthy        #-}
+#else
+{-# LANGUAGE Safe               #-}
+#endif
+----------------------------------------------------------------------------
+-- |
+-- Module      :  Algebra.Lattice.Flat
+-- Copyright   :  (C) 2010-2015 Maximilian Bolingbroke, 2015 Oleg Grenrus
+-- License     :  BSD-3-Clause (see the file LICENSE)
+--
+-- Maintainer  :  Oleg Grenrus <oleg.grenrus@iki.fi>
+--
+----------------------------------------------------------------------------
+module Algebra.Lattice.Flat (
+    Flat(..)
+  ) where
+
+import Prelude ()
+import Prelude.Compat
+
+import Algebra.Lattice
+import Algebra.PartialOrd
+
+import Control.DeepSeq
+import Control.Monad
+import Data.Data
+import Data.Hashable
+import GHC.Generics
+
+--
+-- Flat
+--
+
+-- | Graft a distinct top and bottom onto any type.
+-- The top is the absorbing element for the join.
+-- The bottom is the identity for the join.
+-- Two non-top, non-bottom values join to top, unless they are equal.
+data Flat a = Top
+            | Flat a
+            | Bottom
+  deriving ( Eq, Ord, Show, Read, Data, Typeable, Generic, Functor, Foldable, Traversable
+#if __GLASGOW_HASKELL__ >= 706
+           , Generic1
+#endif
+           )
+
+instance Applicative Flat where
+  pure = return
+  (<*>) = ap
+
+instance Monad Flat where
+  return            = Flat
+  Top >>= _         = Top
+  Bottom >>= _      = Bottom
+  Flat x >>= f  = f x
+
+instance NFData a => NFData (Flat a) where
+  rnf Top          = ()
+  rnf Bottom       = ()
+  rnf (Flat a) = rnf a
+
+instance Hashable a => Hashable (Flat a)
+
+instance Eq a => JoinSemiLattice (Flat a) where
+  Top    \/ _      = Top
+  _      \/ Top    = Top
+  Bottom \/ flat_y = flat_y
+  flat_x \/ Bottom = flat_x
+  Flat x \/ Flat y = if x == y then Flat x else Top
+
+instance Eq a => BoundedJoinSemiLattice (Flat a) where
+  bottom = Bottom
+
+instance Eq a => PartialOrd (Flat a) where
+  leq Bottom Bottom = True
```

Comment:
could we rewrite this so `leq` to be a bit lazier?
```haskell
leq Bottom _ = True
leq _ Top = True
leq Top _ = False
leq _ Bottom = False
leq (Flat x ) (Flat x) = x == y
```

Reply:
Thanks! That's a good idea. I have made it lazier, but I haven't used exactly the code you provided.
The only case where the left hand side determines the result is `leq Bottom _`. Anything after that, and we have to evalute the right to whnf as well. That being the case, my change is as lazy as yours.

I prefer this style because it avoids overlapping patterns, which I find makes the whole pattern match easier to understand, since I can reason about each of the patterns individually. I have also rewritten `\/` to avoid overlapping patterns since I forgot initially.

This style is just my preference though. I don't mind going with the code you provided if you prefer that style.